rpc/src/v1/helpers/errors.rs

```diff
@@ -196,6 +199,53 @@ pub fn no_work_required() -> Error {
 	}
 }
 
+pub fn unavailable_block() -> Error {
+	Error {
+		code: ErrorCode::ServerError(codes::UNSUPPORTED_REQUEST),
+		message: "Block is unavailable".into(),
+		data: None,
+	}
+}
+
+pub fn check_block_number_existence<'a, T, C>(client: &'a C, num: BlockNumber) ->
+	impl Fn(Option<T>) -> RpcResult<Option<T>> + 'a
+	where C: BlockChainClient,
+{
+	move |response| {
+		if response.is_none() {
+			if let BlockNumber::Num(block_number) = num {
+				// tried to fetch block number and got nothing even though the block number is
+				// less than the latest block number
+				if block_number < client.chain_info().best_block_number {
+					return Err(unavailable_block());
+				}
+			}
+		}
+		Ok(response)
+	}
+}
+
+pub fn check_block_gap<'a, T, C>(client: &'a C) -> impl Fn(Option<T>) -> RpcResult<Option<T>> + 'a
+	where C: BlockChainClient,
+{
+	move |response| {
+		if response.is_none() {
+			let BlockChainInfo { ancient_block_hash, first_block_hash, .. } = client.chain_info();
+			// block information was requested, but unfortunately we couldn't find it and there
+			// are gaps in the database ethcore/src/blockchain/blockchain.rs:202
+			if ancient_block_hash.is_some() && first_block_hash.is_some() {
+				return Err(Error {
+					code: ErrorCode::ServerError(codes::UNSUPPORTED_REQUEST),
+					// this error message feels pretty straightforward to me. ¯\_(ツ)_/¯
+					message: "We cannot tell for sure if the thing you requested is not available or we just don't have it".into(),
```

Comment:
Probably better to be explicit about it? "Ancient block sync is still in progress."